pages/authNotAllowed/authNotAllowed.html.js

```diff
@@ -0,0 +1,79 @@
+const { html } = require('@prairielearn/html');
+const { renderEjs } = require('@prairielearn/html-ejs');
+
+function getUrlForProvider(provider) {
+  switch (provider.name) {
+    case 'SAML':
+      return `/pl/auth/institution/${provider.institution_id}/saml/login`;
+    case 'Google':
+      return '/pl/oauth2login';
+    case 'Azure':
+      return '/pl/azure_login';
+    case 'Shibboleth':
+      return '/pl/shibcallback';
+    default:
+      return null;
+  }
+}
+
+function AuthNotAllowed({ institutionAuthnProviders, resLocals }) {
+  // LTI providers are special in that we don't currently allow an LTI session
+  // to be initiated from PrairieLearn. We'll handle them specially.
+  const hasLtiProvider = institutionAuthnProviders.some((p) => p.name === 'LTI');
+  const nonLtiAuthnProviders = institutionAuthnProviders.filter((p) => p.name !== 'LTI');
+  const hasNonLtiProviders = nonLtiAuthnProviders.length > 0;
+
+  // Check if there's a default authentication provider.
+  const defaultAuthnProvider = nonLtiAuthnProviders.find((p) => p.is_default);
+
+  return html`
+    <!DOCTYPE html>
+    <html lang="en">
+      <head>
+        ${renderEjs(__filename, "<%- include('../partials/head') %>", resLocals)}
+      </head>
+      <body>
+        ${renderEjs(__filename, "<%- include('../partials/navbar') %>", {
+          ...resLocals,
+          navbarType: 'unauthenticated',
+        })}
+        <div class="container">
+          <h1>Unsupported authentication method</h1>
+          <p>
+            The authentication method you tried to use is not supported by your institution.
+            ${hasNonLtiProviders ? 'Please use an alternative authentication method.' : ''}
+            ${!hasNonLtiProviders && !hasLtiProvider
+              ? 'Please contact your institution for more information.'
+              : ''}
+          </p>
+          ${defaultAuthnProvider
+            ? html`
+                <div class="mb-3">
+                  <a href="${getUrlForProvider(defaultAuthnProvider)}" class="btn btn-primary">
+                    Sign in with institution single sign-on
+                  </a>
+                </div>
+              `
+            : // If there's no default, show all options (except LTI).
+              nonLtiAuthnProviders.map(
+                (provider) => html`
+                  <div class="mb-3">
+                    <a href="${getUrlForProvider(provider)}" class="btn btn-primary">
+                      Sign in with
+                      ${provider.name === 'SAML' ? 'institution single sign-on' : provider.name}
```

Comment:
Should this also look exactly the same as the buttons on the original login screen? ("U of I" rather than "institution SSO")

Reply:
I think the solution here is going to be the institution dropdown that basically filters the login buttons. I can do that in this PR if you'd like.

Reply:
I'm actually slightly unsure of how this will look with a dropdown. What were you thinking?

Some ideas:
1. A two-part control with a dropdown and a "Login" button. The dropdown starts on "Choose institution" and the button starts out disabled. Selecting an institution from the dropdown enables the button.
2. An immediately activating dropdown, where selecting an entry immediately proceeds to login with that institution.
3. Not a dropdown at all, but instead a vertically scrollable div with buttons saying "Login with U of I", etc. Clicking a button immediately proceeds to login. The scrollable area should show about 5 buttons at a time.
4. Something with a text box to allow searching for the institution, possibly together with a dropdown and a Login button.

Reply:
Whatever we do here, I think we should definitely try and not have buttons that say things like "institution single sign-on" -- I think people are much more like to click "sign on with U of I" or something else with their institutions name on it. A generic "Login" button next to something that says their institution name would also be fine, I think.

Reply:
I had (1) in mind! Do you have any preference one way or another?

Reply:
Re: your other comment, I guess I'm struggling with how to handle the fact that an institution can have multiple allowable providers. Say an institution had both SAML and Google, and they set Google as the default: how do we label the SAML button? Say they don't set a default at all: which button should say "sign in with U of X"? My current design might be optimizing for cases like these that might be unlikely in practice, but I think we still need to handle them.

Reply:
(1) sounds good to me!

Reply:
(especially with a cookie to speed up re-logins)

Reply:
Re: institutions setting google as the default. This all gets a bit messy, I agree. How about this:

For a plain login screen (no query param indicating an institution), we provide:

* All institutions get a "Login with U of X" institution button. The meaning of this button is:
    * If the institution has SAML setup, the institution button goes to SAML login
    * Otherwise, if the institution has a default provider, the institution button goes to the default provider
    * Otherwise, the institution button goes to the first of [Google, Azure] that's enabled
* We always display the "Login with Google/Azure" buttons.

For a login screen with a query param specifying an institution, we display the same buttons as above, but filtered to only display those allowed for the specific institution.

This system means that if you enable SAML but set google as your default, then the default is effectively ignored. I think this is ok because this seems like a weird edge case to me.

Reply:
Discussed with Matt. We're going to avoid any changes to the initial login page for now and focus only on improving the case where the user picks a provider that isn't supported by their institution.